chapter-06/inprogress/suggestions.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import "sort"
+
+// bookSuggestions are the suggested books for a given book.
+// Suggestions are the neighbour-books on the shelves.
+type bookSuggestions map[Book]bookCollection
+
+// bookCollection is a collection of books.
+type bookCollection map[Book]struct{}
+
+// suggestOtherBooks returns the suggestions book from the matching-book.
+func suggestOtherBooks(bookworms []Bookworm) []Bookworm {
+	sb := make(bookSuggestions)
+
+	// Register all books on everyone's shelf.
+	for _, bookworm := range bookworms {
+		for i, book := range bookworm.Books {
+			// Each book on the shelf will be a suggestion for those who have read any book on the shelf.
+			otherBooksOnShelves := listOtherBooksOnShelves(i, bookworm.Books)
+			registerBookSuggestions(sb, book, otherBooksOnShelves)
+		}
+	}
+
+	// Suggest a list of related books to each bookworm.
+	suggestions := make([]Bookworm, len(bookworms))
+	for i, bookworm := range bookworms {
+		suggestions[i] = Bookworm{
+			Name:  bookworm.Name,
+			Books: suggestBooks(sb, bookworm.Books),
+		}
+	}
+
+	return suggestions
+}
+
+// listOtherBooksOnShelves returns the list of my books except the one at the given index.
+func listOtherBooksOnShelves(bookIndexToRemove int, myBooks []Book) []Book {
+	// Initialise the first slice: its capacity is the input slice's capacity reduced by 1, and its starting length is the number of items up to the index to discard.
+	otherBooksOnShelves := make([]Book, bookIndexToRemove, len(myBooks)-1)
+	// Copy the slice of books up to the given index into the initialised index.
+	copy(otherBooksOnShelves, myBooks[:bookIndexToRemove])
+	// Append with the rest of myBooks after the index of the discarded book into the created slice.
+	otherBooksOnShelves = append(otherBooksOnShelves, myBooks[bookIndexToRemove+1:]...)
+
+	return otherBooksOnShelves
+}
+
+// registerBookSuggestions registers the books on the same shelf.
+func registerBookSuggestions(suggestions bookSuggestions, reference Book, otherBooksOnShelves []Book) {
+	for _, book := range otherBooksOnShelves {
+		// Check if this reference has already been added to the map.
+		if suggestions[reference] == nil {
+			// Create a new bookCollection.
+			suggestions[reference] = make(bookCollection)
+		}
+
+		// Fill the associated books for the book reference.
+		suggestions[reference][book] = struct{}{}
+	}
+}
+
+// suggestBooks returns the list of suggested books for a reference book.
+func suggestBooks(suggestions bookSuggestions, myBooks []Book) []Book {
```

Comment:
I would use a different algorithm for the recommendations, mainly to account for similarities between people. e.g.

```
type Recommendation struct {
	Book  Book
	Score float64
}

func recommend(allReaders []Reader, target Reader, n int) []Recommendation {
	read := newSet(target.Books...)

	recommendations := map[Book]float64{}
	for _, reader := range allReaders {
		if reader.Name == target.Name {
			continue
		}

		// find the similarity between the readers
		var similarity float64
		for _, book := range reader.Books {
			if read.Contains(book) {
				similarity++
			}
			// you could also later extend to liked and dislike score
		}
		if similarity == 0 {
			continue
		}

		// this adjust the book scoring, such that it doesn't overpower all other
		// suggestions when one person has too many similarities.
		score := math.Log(similarity) + 1
		for _, book := range readers.Books {
			if !read.Contains(book) {
				recommendations[book] += score
			}
		}
	}

	// TODO: sort by score
	// TODO: only output a certain amount of recommendations
}
```

You can of course, introduce some similar scoring into per-book recommendations as well, but might be more complicated.

Of course, this approach could have a performance problem with large number of readers.

Reply:
We will do it as an exercice and add both versions.